### PR TITLE
Import data from s3

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: config/pmd/ruleset.xml [hunk 1]
```diff
@@ -70,7 +70,10 @@
   </rule>
 
 
-  <rule ref="category/java/performance.xml"/>
+  <rule ref="category/java/performance.xml">
+    <!-- https://softwareengineering.stackexchange.com/questions/149563/should-we-avoid-object-creation-in-java -->
```
Comment (axthosarouris) [inline config/pmd/ruleset.xml:74]: Creating short-lived objects is not a performance issue in Java anymore.

File: data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    private static final int MAX_ATTEMPTS = 10;
+    private final AmazonDynamoDB dynamoClient;
+    private S3Driver s3Driver;
+    private String tableName;
+    private static final Logger logger = LoggerFactory.getLogger(DataImportHandler.class);
+
+    @JacocoGenerated
+    public DataImportHandler() {
+        this(null, defaultDynamoClient());
```
Comment (janbj) [inline data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java:41]: Why is the first argument null?

**Review by janbj — APPROVED** [review]

File: data-import-from-s3/src/main/java/no/unit/nva/dataimport/ImportRequest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    private String folderPath;
+    @JsonProperty("table")
+    private String table;
+
+    @JacocoGenerated
+    public ImportRequest() {
+
+    }
+
+    public ImportRequest(String bucket,
```
Comment (axthosarouris) [inline data-import-from-s3/src/main/java/no/unit/nva/dataimport/ImportRequest.java:21]: @JsonCreator does not seem to work with AWS's serialiser for some reason. Added default serializer.

File: data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                   .collect(Collectors.groupingBy(Entry::getKey))
+                   .values()
+                   .stream()
+                   .flatMap(this::mergeLists)
+                   .collect(Collectors.toMap(SimpleEntry::getKey, SimpleEntry::getValue));
+    }
+
+    private <T> Stream<SimpleEntry<String, List<T>>> mergeLists(List<Entry<String, List<T>>> entries) {
+        List<T> values = entries.stream().flatMap(e -> e.getValue().stream()).collect(Collectors.toList());
+        return attempt(() -> entries.get(0).getKey())
```
Comment (brinxmat) [inline data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java:170]: What is 0 here?

File: data-import-from-s3/src/main/java/no/unit/nva/dataimport/S3Reader.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import software.amazon.ion.IonReader;
+import software.amazon.ion.IonWriter;
+import software.amazon.ion.system.IonReaderBuilder;
+import software.amazon.ion.system.IonTextWriterBuilder;
+
+class S3Reader {
+
+    public static final String PK0 = DatabaseConstants.PRIMARY_KEY_PARTITION_KEY_NAME;
+    // Looking for Strings '<end_of_previous_object>}<possible_white_space>{"Item":{"PK0"'
+    public static final String CONSECUTIVE_JSON_OBJECTS = "(})\\s*(\\{\"Item\":\\{\"" + PK0 + "\")";
```
Comment (brinxmat) [inline data-import-from-s3/src/main/java/no/unit/nva/dataimport/S3Reader.java:22]: Regex is the sign of a bad life, one where data is badly formatted before it reaches you…

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        S3ToDynamoImporter s3ToDynamoImporter = new S3ToDynamoImporter(dynamoClient, s3Driver, tableName, filename);
+        List<BatchWriteItemResult> results = attempt(s3ToDynamoImporter::insertFileToDynamo).orElseThrow();
+        BatchWriteItemResult result = collectResults(results);
+        if (itemsFailedToBeInserted(result)) {
+            throw new BatchInsertionFailureException(result);
+        }
+        return result;
+    }
+
+    private boolean itemsFailedToBeInserted(BatchWriteItemResult result) {
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply data-import-from-s3/src/main/java/no/unit/nva/dataimport/DataImportHandler.java:112]: BatchWriteItemResult is an AWS class

**Review by brinxmat — APPROVED** [review]

**Pull request merged**